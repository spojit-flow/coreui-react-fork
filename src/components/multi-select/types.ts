export type Option = {
  custom?: boolean
  disabled?: boolean
  selected?: boolean
  text: string // TODO: change to label in v5
  value: number | string
  [key: string]: number | string | any
}

export type OptionsGroup = {
  label: string
  options?: Option[]
  [key: string]: number | string | any
}

export type SelectedOption = {
  disabled?: boolean
  text: string
  value: number | string
  [key: string]: number | string | any
}
