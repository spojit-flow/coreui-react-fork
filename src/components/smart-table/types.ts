import { ReactNode } from 'react'

import { CTableDataCellProps } from '../table/CTableDataCell'
import { CTableHeaderCellProps } from '../table/CTableHeaderCell'
import { CTableRowProps } from '../table/CTableRow'

export type Column = {
  children?: Column[]
  filter?: boolean | ((values: any[], onChange: (value: any) => void) => ReactNode)
  group?: string
  key: string
  label?: string
  sorter?: boolean
  _style?: any
  _props?: CTableHeaderCellProps
}

export type ColumnFilter = {
  lazy?: boolean
  external?: boolean
}

export type ColumnFilterValue = {
  [key: string]: any
}

export type FooterItem = {
  label?: string
  _props?: CTableDataCellProps
}

export type Group = {
  children?: Group[] | Column[]
  colspan?: number
  deep?: number
  group?: string
  key: string
  label?: string
  _style?: any
  _props?: CTableHeaderCellProps
}

export type Item = {
  [key: string]: number | string | any
  _cellProps?: any
  _props?: CTableRowProps
  _selected?: boolean
}

export type ItemsPerPageSelect = {
  external?: boolean
  label?: string
  values?: Array<number>
}

export type Pagination = {
  external?: boolean
}

export type Sorter = {
  resetable?: boolean
  external?: boolean
}

export type ScopedColumns = {
  [key: string]: any
  details?: (a: Item, b: number) => ReactNode
}

export type SorterValue = {
  column?: string
  state?: number | string
}

export type TableFilter = {
  lazy?: boolean
  external?: boolean
}
