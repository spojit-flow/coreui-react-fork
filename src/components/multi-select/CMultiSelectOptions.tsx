import React, { forwardRef, HTMLAttributes, ReactNode } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { CElementCover } from '../element-cover'
import { CVirtualScroller } from '../virtual-scroller'

import { getNextSibling, getPreviousSibling } from './utils'
import type { Option, OptionsGroup } from './types'

export interface CMultiSelectOptionsProps extends HTMLAttributes<HTMLDivElement> {
  handleOptionOnClick?: (option: Option) => void
  loading?: boolean
  options: (Option | OptionsGroup)[]
  optionsMaxHeight?: number | string
  optionsStyle?: 'checkbox' | 'text'
  optionsTemplate?: (option: Option) => ReactNode
  optionsGroupsTemplate?: (option: OptionsGroup) => ReactNode
  searchNoResultsLabel?: string | ReactNode
  selected: Option[]
  virtualScroller?: boolean
  visibleItems?: number
}

export const CMultiSelectOptions = forwardRef<HTMLDivElement, CMultiSelectOptionsProps>(
  (
    {
      handleOptionOnClick,
      loading,
      options,
      optionsMaxHeight,
      optionsStyle,
      optionsTemplate,
      optionsGroupsTemplate,
      searchNoResultsLabel,
      selected,
      virtualScroller,
      visibleItems = 10,
    },
    ref,
  ) => {
    const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>, option: Option) => {
      if (event.code === 'Space' || event.key === 'Enter') {
        event.preventDefault()
        handleOptionOnClick && handleOptionOnClick(option)
      }

      if (event.key === 'Down' || event.key === 'ArrowDown') {
        event.preventDefault()
        const target = event.target as HTMLElement
        const next = getNextSibling(target, '.form-multi-select-option')

        next && (next as HTMLElement).focus()
      }

      if (event.key === 'Up' || event.key === 'ArrowUp') {
        event.preventDefault()
        const target = event.target as HTMLElement
        const prev = getPreviousSibling(target, '.form-multi-select-option')

        prev && (prev as HTMLElement).focus()
      }
    }

    const createOptions = (options: (Option | OptionsGroup)[]): JSX.Element | JSX.Element[] =>
      options.length > 0 ? (
        options.map((option: Option | OptionsGroup, index: number) =>
          'value' in option ? (
            <div
              className={classNames('form-multi-select-option', {
                'form-multi-select-option-with-checkbox': optionsStyle === 'checkbox',
                'form-multi-selected': selected.some((_option) => _option.value === option.value),
                disabled: option.disabled,
              })}
              key={index}
              onClick={() => handleOptionOnClick && handleOptionOnClick(option as Option)}
              onKeyDown={(event) => handleKeyDown(event, option as Option)}
              tabIndex={0}
            >
              {optionsTemplate ? optionsTemplate(option as Option) : option.text}
            </div>
          ) : (
            <div className="form-multi-select-optgroup-label" key={index}>
              {optionsGroupsTemplate ? optionsGroupsTemplate(option as OptionsGroup) : option.label}
            </div>
          ),
        )
      ) : (
        <div className="form-multi-select-options-empty">{searchNoResultsLabel}</div>
      )

    return (
      <>
        {virtualScroller ? (
          <CVirtualScroller
            className="form-multi-select-options"
            visibleItems={visibleItems}
            ref={ref}
          >
            {createOptions(options)}
          </CVirtualScroller>
        ) : (
          <div
            className="form-multi-select-options"
            {...(optionsMaxHeight !== 'auto' && {
              style: { maxHeight: optionsMaxHeight, overflow: 'scroll' },
            })}
            ref={ref}
          >
            {createOptions(options)}
          </div>
        )}
        {loading && <CElementCover />}
      </>
    )
  },
)

CMultiSelectOptions.propTypes = {
  handleOptionOnClick: PropTypes.func,
  loading: PropTypes.bool,
  options: PropTypes.array.isRequired,
  optionsMaxHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  optionsStyle: PropTypes.oneOf(['checkbox', 'text']),
  optionsTemplate: PropTypes.func,
  optionsGroupsTemplate: PropTypes.func,
  searchNoResultsLabel: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  virtualScroller: PropTypes.bool,
  visibleItems: PropTypes.number,
}

CMultiSelectOptions.displayName = 'CMultiSelectOptions'
