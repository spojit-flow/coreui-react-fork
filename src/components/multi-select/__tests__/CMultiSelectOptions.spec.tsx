import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { CMultiSelectOptions } from '../CMultiSelectOptions'
import { Option } from '../types'

const options: Option[] = [
  {
    text: 'text1',
    value: 'value1',
    selected: true,
    disabled: true,
  },
  {
    text: 'text3',
    value: 'value3',
  },
]

test('loads and displays CMultiSelectOptions component', async () => {
  const { container } = render(
    <CMultiSelectOptions options={[]} searchNoResultsLabel="some no result label" selected={[]} />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('form-multi-select-options')
  if (container.firstChild === null) {
    expect(true).toBe(false)
  } else {
    expect(container.firstChild.firstChild).toHaveClass('form-multi-select-options-empty')
    expect(container.firstChild.firstChild).toHaveTextContent('some no result label')
  }
})

test('CMultiSelectOptions customize', async () => {
  const { container } = render(
    <CMultiSelectOptions
      options={options}
      optionsMaxHeight="200"
      optionsStyle="checkbox"
      searchNoResultsLabel="some - searchNoResultsLabel"
      selected={[]}
    />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('form-multi-select-options')
  expect(container.firstChild).toHaveAttribute('style', 'max-height: 200; overflow: scroll;')
  if (container.firstChild === null) {
    expect(true).toBe(false)
  } else {
    expect(container.firstChild.firstChild).toHaveClass('form-multi-select-option')
    expect(container.firstChild.firstChild).toHaveClass('form-multi-select-option-with-checkbox')
    expect(container.firstChild.firstChild).toHaveClass('disabled')
    expect(container.firstChild.firstChild).toHaveTextContent('text1')
    expect(container.firstChild.lastChild).toHaveClass('form-multi-select-option')
    expect(container.firstChild.lastChild).toHaveClass('form-multi-select-option-with-checkbox')
    expect(container.firstChild.lastChild).toHaveTextContent('text3')
  }
})

test('CMultiSelectOptions handle on click', async () => {
  const onClick = jest.fn()
  render(<CMultiSelectOptions options={options} handleOptionOnClick={onClick} selected={[]} />)
  expect(onClick).toHaveBeenCalledTimes(0)
  const option = document.querySelector('.form-multi-select-option')
  if (option !== null) {
    fireEvent.click(option)
  }
  expect(onClick).toHaveBeenCalledTimes(1)
})
