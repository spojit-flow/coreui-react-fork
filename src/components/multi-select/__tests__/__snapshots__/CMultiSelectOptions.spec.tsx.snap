// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMultiSelectOptions customize 1`] = `
<div>
  <div
    class="form-multi-select-options"
    style="max-height: 200; overflow: scroll;"
  >
    <div
      class="form-multi-select-option form-multi-select-option-with-checkbox disabled"
      tabindex="0"
    >
      text1
    </div>
    <div
      class="form-multi-select-option form-multi-select-option-with-checkbox"
      tabindex="0"
    >
      text3
    </div>
  </div>
</div>
`;

exports[`loads and displays CMultiSelectOptions component 1`] = `
<div>
  <div
    class="form-multi-select-options"
    style="overflow: scroll;"
  >
    <div
      class="form-multi-select-options-empty"
    >
      some no result label
    </div>
  </div>
</div>
`;
